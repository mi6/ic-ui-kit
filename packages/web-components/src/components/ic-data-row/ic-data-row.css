@import "../../global/normalize.css";

:host {
  position: relative;
  display: block;
}

.data {
  display: flex;
  align-items: center;
}

.text-cells {
  display: flex;
  flex-grow: 1;
  min-width: 0;
}

.label {
  width: 12.5rem;
  min-width: 12.5rem;
  margin-right: var(--ic-space-md);
  color: var(--ic-color-tertiary-text);
}

.value {
  flex-grow: 1;
}

slot[name="value"]::slotted(ic-text-field[readonly][hide-label]) {
  margin-top: calc(var(--ic-space-xs) * -1);
}

slot[name="value"]::slotted(ic-text-field[small][readonly][hide-label]) {
  margin-top: calc(var(--ic-space-xxs) * -1);
}

slot[name="value"]::slotted(ic-text-field[readonly][hide-label][rows]) {
  margin-top: calc(var(--ic-space-xxs) * -1 - var(--ic-space-xxxs));
}

.end-component {
  width: fit-content;
  margin-left: 3.5rem;
}

.divider {
  margin-top: var(--ic-space-md);
  height: var(--ic-space-1px);
  background-color: var(--ic-architectural-300);
}

:host(.small) .divider {
  margin-top: var(--ic-space-xs);
}

:host(.breakpoint-medium) .label {
  width: 10rem;
  min-width: 10rem;
}

:host(.breakpoint-xs) .text-cells {
  flex-direction: column;
}

:host(.breakpoint-xs) .label {
  width: 8rem;
  margin-bottom: var(--ic-space-xs);
}

:host(.breakpoint-xs) .value {
  max-width: 95%;
}

:host(.breakpoint-xs) .end-component {
  margin-left: 0;
}

@media (forced-colors: active) {
  .divider {
    background-color: canvastext;
  }
}
