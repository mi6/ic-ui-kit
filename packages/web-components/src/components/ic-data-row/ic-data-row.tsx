import { Component, Element, Host, Prop, h, State } from "@stencil/core";
import {
  checkResizeObserver,
  DEVICE_SIZES,
  getCurrentDeviceSize,
  isSlotUsed,
  slotHasContent,
} from "../../utils/helpers";

/**
 * @slot label - Content will be rendered in the leftmost cell.
 * @slot value - Content will be rendered to the right of the label.
 * @slot end-component - Content will be displayed in the rightmost cell.
 */
@Component({
  tag: "ic-data-row",
  styleUrl: "ic-data-row.css",
  shadow: true,
})
export class DataRow {
  private hasEndComponent: boolean = false;
  private resizeObserver: ResizeObserver = null;

  @Element() el: HTMLIcDataRowElement;

  @State() deviceSize: number = DEVICE_SIZES.XL;
  @State() entitySize: "xl" | "m" | "xs";

  /**
   * The label in the leftmost cell of the row.
   */
  @Prop() label: string;

  /**
   * If `true`, the small styling will be applied to the row. This is automatically applied if small is set on the parent data heading.
   */
  @Prop() small: boolean = false;

  /**
   * The value of the middle (right if no end-component supplied) cell of the row.
   */
  @Prop() value: string;

  disconnectedCallback(): void {
    if (this.resizeObserver !== null) {
      this.resizeObserver.disconnect();
    }
  }

  componentWillLoad(): void {
    this.deviceSize = getCurrentDeviceSize();
    this.hasEndComponent = slotHasContent(this.el, "end-component");
    this.checkLabelAbove();
  }

  componentDidLoad(): void {
    checkResizeObserver(this.runResizeObserver);
    if (this.hasEndComponent) {
      this.labelEndComponent();
    }
  }

  private runResizeObserver = () => {
    this.resizeObserver = new ResizeObserver(() => {
      this.checkLabelAbove();
    });

    this.resizeObserver.observe(this.el);
  };

  private checkLabelAbove() {
    const rowSize = this.el.shadowRoot.querySelector(".data")?.clientWidth;
    if (rowSize) {
      if (rowSize + 46 < DEVICE_SIZES.S) {
        this.entitySize = "xs";
      } else if (rowSize + 46 < DEVICE_SIZES.M) {
        this.entitySize = "m";
      } else {
        this.entitySize = "xl";
      }
    }
  }

  /**
   * Renders the label either as a ic-typography or slot. Slotted content takes precedence.
   * @param label string - label of value
   * @returns HTMLDivElement - returns label as slot or ic-typography with label as textContent
   */
  private renderLabel = (label: string) => {
    if (isSlotUsed(this.el, "label")) {
      return (
        <div class="label">
          <slot name="label"></slot>
        </div>
      );
    } else if (label) {
      return (
        <div class="label">
          <ic-typography
            variant={this.entitySize === "xs" ? "label" : "subtitle-large"}
          >
            {label}
          </ic-typography>
        </div>
      );
    }

    return null;
  };

  private labelEndComponent(): void {
    const component = this.el.shadowRoot.querySelectorAll(
      "slot[name=end-component]"
    );

    component.forEach((child: HTMLElement) =>
      child.setAttribute("aria-label", "for " + this.label + " row")
    );
  }

  render() {
    const { label, value, small } = this;

    return (
      <Host
        class={{
          ["small"]: small,
          ["breakpoint-medium"]: this.entitySize === "m",
          ["breakpoint-xs"]: this.entitySize === "xs",
        }}
        role="listitem"
      >
        <div class="data">
          <div class="text-cells">
            {this.renderLabel(label)}
            <div class="value">
              <slot name="value">
                <ic-typography variant="body">{value}</ic-typography>
              </slot>
            </div>
          </div>
          {this.hasEndComponent && (
            <div class="end-component">
              <div role="cell">
                <slot name="end-component"></slot>
              </div>
            </div>
          )}
        </div>
        <div class="divider" />
      </Host>
    );
  }
}
