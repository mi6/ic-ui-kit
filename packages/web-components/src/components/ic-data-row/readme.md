# ic-data-row



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                                                                                                        | Type      | Default     |
| -------- | --------- | ---------------------------------------------------------------------------------------------------------------------------------- | --------- | ----------- |
| `label`  | `label`   | The label in the leftmost cell of the row.                                                                                         | `string`  | `undefined` |
| `small`  | `small`   | If `true`, the small styling will be applied to the row. This is automatically applied if small is set on the parent data heading. | `boolean` | `false`     |
| `value`  | `value`   | The value of the middle (right if no end-component supplied) cell of the row.                                                      | `string`  | `undefined` |


## Slots

| Slot              | Description                                         |
| ----------------- | --------------------------------------------------- |
| `"end-component"` | Content will be displayed in the rightmost cell.    |
| `"label"`         | Content will be rendered in the leftmost cell.      |
| `"value"`         | Content will be rendered to the right of the label. |


## Dependencies

### Depends on

- [ic-typography](../ic-typography)

### Graph
```mermaid
graph TD;
  ic-data-row --> ic-typography
  style ic-data-row fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


