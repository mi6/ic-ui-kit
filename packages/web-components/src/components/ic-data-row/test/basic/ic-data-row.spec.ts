import { newSpecPage } from "@stencil/core/testing";
import { DataRow } from "../../ic-data-row";

describe("ic-data-row", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label" value="value"></ic-data-row>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted content in the label slot", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row value="value"><ic-typography slot="label">Label</ic-typography></ic-data-row>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should not render label element if no label provided", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row value="value"></ic-data-row>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted content in the value slot", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label"><ic-status-tag variant="success" label="success" slot="value"></ic-status-tag></ic-data-row>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render slotted content in the end-component slot", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label" value="test value"><ic-status-tag variant="success" label="success" slot="end-component"></ic-status-tag></ic-data-row>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render the label variant of typography when entity size is xs", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label"><ic-status-tag variant="success" label="success" slot="value"></ic-status-tag></ic-data-row>`,
    });

    page.rootInstance.entitySize = "xs";
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  it("should call runResizeObserver", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label" value="value"></ic-data-row>`,
    });

    await page.rootInstance.runResizeObserver();
    page.waitForChanges();

    const resize = new ResizeObserver(() => {
      page.rootInstance.checkLabelAbove();
    });

    expect(page.rootInstance.resizeObserver).toBe(resize);

    page.setContent("");
  });

  it("should change entity size depending on screen size", async () => {
    const page = await newSpecPage({
      components: [DataRow],
      html: `<ic-data-row label="label" value="value"></ic-data-row>`,
    });

    Object.defineProperty(
      page.root.shadowRoot.querySelector(".data"),
      "clientWidth",
      {
        value: 200,
      }
    );

    page.waitForChanges();

    page.rootInstance.checkLabelAbove();

    expect(page.rootInstance.entitySize).toBe("xs");
  });
});
