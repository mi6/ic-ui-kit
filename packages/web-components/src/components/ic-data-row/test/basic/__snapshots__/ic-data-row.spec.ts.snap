// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-data-row should not render label element if no label provided 1`] = `
<ic-data-row role="listitem" value="value">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="value">
          <slot name="value">
            <ic-typography variant="body">
              value
            </ic-typography>
          </slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
</ic-data-row>
`;

exports[`ic-data-row should render 1`] = `
<ic-data-row label="label" role="listitem" value="value">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="label">
          <ic-typography variant="subtitle-large">
            label
          </ic-typography>
        </div>
        <div class="value">
          <slot name="value">
            <ic-typography variant="body">
              value
            </ic-typography>
          </slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
</ic-data-row>
`;

exports[`ic-data-row should render slotted content in the end-component slot 1`] = `
<ic-data-row label="label" role="listitem" value="test value">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="label">
          <ic-typography variant="subtitle-large">
            label
          </ic-typography>
        </div>
        <div class="value">
          <slot name="value">
            <ic-typography variant="body">
              test value
            </ic-typography>
          </slot>
        </div>
      </div>
      <div class="end-component">
        <div role="cell">
          <slot aria-label="for label row" name="end-component"></slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
  <ic-status-tag label="success" slot="end-component" variant="success"></ic-status-tag>
</ic-data-row>
`;

exports[`ic-data-row should render slotted content in the label slot 1`] = `
<ic-data-row role="listitem" value="value">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="label">
          <slot name="label"></slot>
        </div>
        <div class="value">
          <slot name="value">
            <ic-typography variant="body">
              value
            </ic-typography>
          </slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
  <ic-typography slot="label">
    Label
  </ic-typography>
</ic-data-row>
`;

exports[`ic-data-row should render slotted content in the value slot 1`] = `
<ic-data-row label="label" role="listitem">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="label">
          <ic-typography variant="subtitle-large">
            label
          </ic-typography>
        </div>
        <div class="value">
          <slot name="value">
            <ic-typography variant="body"></ic-typography>
          </slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
  <ic-status-tag label="success" slot="value" variant="success"></ic-status-tag>
</ic-data-row>
`;

exports[`ic-data-row should render the label variant of typography when entity size is xs 1`] = `
<ic-data-row class="breakpoint-xs" label="label" role="listitem">
  <mock:shadow-root>
    <div class="data">
      <div class="text-cells">
        <div class="label">
          <ic-typography variant="label">
            label
          </ic-typography>
        </div>
        <div class="value">
          <slot name="value">
            <ic-typography variant="body"></ic-typography>
          </slot>
        </div>
      </div>
    </div>
    <div class="divider"></div>
  </mock:shadow-root>
  <ic-status-tag label="success" slot="value" variant="success"></ic-status-tag>
</ic-data-row>
`;
